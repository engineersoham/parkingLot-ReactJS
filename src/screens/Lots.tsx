/** @format */

import React, { FC, useEffect, useState } from "react";
import {
	Button,
	FlatList,
	Modal,
	SafeAreaView,
	Text,
	TextInput,
	TouchableOpacity,
	View,
} from "react-native";
import { Snackbar } from "react-native-paper";
import { style } from "../styles/screens";
import { lotDetailsType } from "../types/components";
import { lotsPropsType } from "../types/screens";

const Lots: FC<lotsPropsType> = (props) => {
	const [lotsList, setLotsList] = useState<lotDetailsType[]>([]);
	const [currentLot, setCurrentLot] = useState<number>(0);
	const [freeLotsList, setFreeLotsList] = useState<lotDetailsType[]>(lotsList);
	const [reg, setReg] = useState<string>("");
	const [showAddModal, setShowAddModal] = useState<boolean>(false);
	const [showRemoveModal, setShowRemoveModal] = useState<boolean>(false);
	const [showSnack, setShowSnack] = useState<boolean>(false);

	const [hrs, setHrs] = useState<number>(0);
	const [amnt, setAmnt] = useState<number>(0);

	useEffect(() => {
		drawLots();
	}, []);

	useEffect(() => {
		setFreeLotsList(lotsList.filter((lot) => lot.free));
	}, [lotsList]);

	function drawLots() {
		let lotsArray = [];
		for (let i = 0; i < props.route.params.lots; i++) {
			lotsArray.push({
				id: i,
				reg: "",
				free: true,
				start: new Date(0, 0, 0),
			});
		}
		setLotsList(lotsArray);
	}

	function getRandom() {
		const randomNum = Math.floor(Math.random() * freeLotsList.length);
		setCurrentLot(freeLotsList[randomNum].id);
	}

	function handleAdd(random: boolean) {
		setReg("");
		if (freeLotsList.length > 0) {
			if (random) {
				getRandom();
			}
			if (currentLot >= 0) {
				setShowAddModal(true);
			}
		} else {
			setShowSnack(true);
			setTimeout(() => {
				setShowSnack(false);
			}, 2000);
		}
	}

	function handleRemove() {
		!lotsList[currentLot].free && setShowRemoveModal(true);
	}

	function calculateHrsAmt() {
		const timeDiffms = Math.abs(
			lotsList[currentLot].start.getTime() - new Date().getTime()
		);
		const timeDiffHrs = Math.floor(timeDiffms / (1000 * 60 * 60));
		setHrs(timeDiffHrs);

		if (timeDiffHrs <= 2) {
			setAmnt(10);
		} else {
			setAmnt(10 + (timeDiffHrs - 2) * 10);
		}
	}

	return (
		<SafeAreaView testID="lots" style={style.container2}>
			{/* Add Modal */}
			<Modal visible={showAddModal} animationType="slide">
				<View style={style.modal}>
					<Text>Add Vehicle to P{currentLot}</Text>
					<TextInput
						placeholder="Enterv reg. number"
						placeholderTextColor={"grey"}
						onChangeText={(text) => {
							setReg(text);
						}}
						style={style.input}
					/>

					<View style={style.buttonRow}>
						<Button
							disabled={reg.length == 0}
							title="Add"
							onPress={() => {
								if (reg.length) {
									setLotsList(
										lotsList.map((lot) => {
											return lot.id == currentLot
												? {
														...lot,
														free: false,
														reg: reg,
														start: new Date(),
												  }
												: lot;
										})
									);
									setShowAddModal(false);
								}
							}}
						/>
						<Button
							title="Cancel"
							onPress={() => {
								setShowAddModal(false);
							}}
						/>
					</View>
				</View>
			</Modal>
			{/* Remove Modal */}
			<Modal
				visible={showRemoveModal}
				onShow={() => {
					calculateHrsAmt();
				}}
				animationType="slide"
			>
				<View style={style.modal}>
					<Text>Pay and Remove Vehicle from P{currentLot}</Text>
					<Text>Total hours: {hrs}</Text>
					<Text>Total Amount:{amnt}</Text>

					<View style={style.buttonRow}>
						<Button
							title="Remove"
							onPress={() => {
								setLotsList(
									lotsList.map((lot) => {
										return lot.id == currentLot
											? {
													...lot,
													free: true,
													reg: "",
													start: new Date(0, 0, 0),
											  }
											: lot;
									})
								);
								setAmnt(0);
								setHrs(0);
								setShowRemoveModal(false);
							}}
						/>
						<Button
							title="Cancel"
							onPress={() => {
								setAmnt(0);
								setHrs(0);
								setShowRemoveModal(false);
							}}
						/>
					</View>
				</View>
			</Modal>

			<Snackbar
				visible={showSnack}
				onDismiss={() => setShowSnack(false)}
				style={style.snack}
			>
				<Text style={style.snackText}>the parking is full</Text>
			</Snackbar>

			<TouchableOpacity
				onPress={() => handleAdd(true)}
				style={style.parkingArea}
			>
				<FlatList
					data={lotsList}
					renderItem={({ item }) => (
						<TouchableOpacity
							onPress={() => {
								setCurrentLot(item.id);
								item.free ? handleAdd(false) : handleRemove();
							}}
						>
							<View
								style={{
									...style.item,
									backgroundColor: item.free ? "green" : "red",
								}}
							>
								<Text style={style.itemText}>P{item.id}</Text>
								<Text style={style.itemText}>
									{item.free ? "Free" : `Occupied by ${item.reg}`}
								</Text>
							</View>
						</TouchableOpacity>
					)}
				/>
			</TouchableOpacity>
		</SafeAreaView>
	);
};

export default Lots;
