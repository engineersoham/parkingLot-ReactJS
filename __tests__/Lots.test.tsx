/** @format */

import { render } from "@testing-library/react-native";
import Lots from "../src/screens/Lots";

jest.mock("@react-navigation/native", () => {
	return {
		useNavigation: jest.fn(),
	};
});

const route = {
	params: {
		lots: 2,
	},
};

describe("Lots", () => {
	it("renders without crash", () => {
		render(<Lots route={route} />);
	});

	it("matches snapshot", () => {
		const tree = render(<Lots route={route} />);
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("has component with req TestId", () => {
		const tree = render(<Lots route={route} />);
		expect(tree.findByTestId("lots")).toBeTruthy();
	});

	it("has 3 children", () => {
		const tree = render(<Lots route={route} />);
		//@ts-ignore
		expect(tree.toJSON().children.length).toBe(3);
	});
});
