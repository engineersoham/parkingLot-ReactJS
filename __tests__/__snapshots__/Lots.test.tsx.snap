// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lots matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 10,
    }
  }
  testID="lots"
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "top": "30%",
        }
      }
    >
      <Text>
        Add Vehicle to P
        0
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Enterv reg. number"
        placeholderTextColor="grey"
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "grey",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 10,
            "padding": 5,
          }
        }
        underlineColorAndroid="transparent"
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
            "width": "80%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
                Object {},
              ]
            }
          >
            <Text
              disabled={true}
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#cdcdcd",
                  },
                ]
              }
            >
              Add
            </Text>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onShow={[Function]}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "top": "30%",
        }
      }
    >
      <Text>
        Pay and Remove Vehicle from P
        0
      </Text>
      <Text>
        Total hours: 
        0
      </Text>
      <Text>
        Total Amount:
        0
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
            "width": "80%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Remove
            </Text>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "padding": "10%",
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "free": true,
            "id": 0,
            "reg": "",
            "start": 1899-12-30T18:38:50.000Z,
          },
          Object {
            "free": true,
            "id": 1,
            "reg": "",
            "start": 1899-12-30T18:38:50.000Z,
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "green",
                  "borderRadius": 10,
                  "flexDirection": "row",
                  "justifyContent": "space-around",
                  "marginHorizontal": 10,
                  "marginVertical": 15,
                  "padding": 20,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                P
                0
              </Text>
              <Text
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                Free
              </Text>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "green",
                  "borderRadius": 10,
                  "flexDirection": "row",
                  "justifyContent": "space-around",
                  "marginHorizontal": 10,
                  "marginVertical": 15,
                  "padding": 20,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                P
                1
              </Text>
              <Text
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                Free
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
